hazardous/metrics/_concordance_index.py

```diff
@@ -0,0 +1,409 @@
+
+    Returns
+    -------
+    cindex: list of float
+        Value of the concordance index
+
+    n_pairs_a: list of int
+        Number of comparable pairs with T_i <= T_j (type A) without ties for D_j != 0
+        those are the only comparable pairs for survival without competing events.
+
+    n_concordant_pairs_a: list of int
+        Number of concordant pairs among A pairs without ties.
+
+    n_ties_times_a: list of int
+        Number of tied pairs of type A with D_i = D_j = 1.
+
+    n_ties_pred_a: list of int
+        Number of pairs of type A with np.abs(y_pred_i - y_pred_j) <= tied_tol
+
+    n_pairs_b: list of int
+        Number of comparable pairs with T_i >= T_j where j has
+        a competing event (type B).
+        0 if there are no competing events.
+
+    n_concordant_pairs_b: list of int
+        Number of concordant pairs among B pairs without ties.
+
+    n_ties_pred_b: list of int
+        Number of pairs of type B with np.abs(y_pred_i - y_pred_j) <= tied_tol.
+    """
+    if y_pred.ndim != 2:
+        raise ValueError(
+            "y_pred dimension must be 2, with shape (n_samples, n_time_grid), "
+            f"got {y_pred.ndim=}."
+        )
+    if y_test.shape[0] != y_pred.shape[0]:
+        raise ValueError(
+            "y_test and y_pred must be the same length, got:"
+            f"{y_test.shape[0]=} and {y_pred.shape[0]=}."
+        )
+    if time_grid is not None and len(time_grid) != y_pred.shape[1]:
+        raise ValueError(
+            "y_pred must have the same number of columns as the length of time_grid, "
+            f"got: {y_test.shape[1]=} and {len(time_grid)=}."
+        )
+
+    if ipcw_estimator is None:
+        if y_train is not None:
+            warnings.warn(
+                "y_train passed but ipcw_estimator is set to None, "
+                "therefore y_train won't be used. Set y_train=None "
+                "to silence this warning."
+            )
+        ipcw = np.ones(y_test["event"].shape[0])
+
+    else:
+        if y_train is None:
+            # Raising here since the error raised by the IPCW estimator doesn't
+            # help the user understand that y_train is missing.
+            raise ValueError(
+                "ipcw_estimator is set, but y_train is None. "
+                "Set y_train to fix this error."
+            )
+        # TODO: add cox option
+        ipcw_estimator_ = KaplanMeierIPCW().fit(y_train)
+        ipcw = ipcw_estimator_.compute_ipcw_at(
+            y_test["duration"]
+        )  # shape: (n_samples_test,)
+
+    if taus is None:
+        t_min, t_max = 0, y_test["duration"].max()
+        taus = [t_max]
+        if time_grid is None:
+            time_grid = np.linspace(t_min, t_max, num=y_pred.shape[1])
+    else:
+        if time_grid is None:
+            raise ValueError("When 'taus' is set, 'time_grid' must also be set.")
+    taus = np.atleast_1d(taus)
+
+    c_index_report = defaultdict(list)
+    c_index_report["taus"] = taus
```

Comment:
So here using a defaultdict of list is handy because we can accumulate items in a dict of list, a few lines below:
```python
        for metric_name, metric_value in c_index_report_tau.items():
            c_index_report[metric_name].append(metric_value)
```
Otherwise, we would have to create a list for each key, which would introduce some boilerplate. WDYT?

Reply:
ah indeed, I missed that part. Keeping the `defaultdict(list)` sounds good.